src/app/static/src/app/apiService.ts

```diff
@@ -1,47 +1,122 @@
+    | ModelRunResultResponse
+    | InitialiseModelRunResponse
+    | InternalResponse
+    | Boolean
+
+declare var appUrl: string;
 
-type ResponseData = ValidateInputResponse | ModelRunResultResponse | InitialiseModelRunResponse | InternalResponse | null
+export interface API<S, E> {
 
-export class APIService {
+    withError: (type: E) => API<S, E>
+    withSuccess: (type: S) => API<S, E>
+    ignoreErrors: () => API<S, E>
+
+    postAndReturn<T extends ResponseData>(url: string, data: any): Promise<void | T>
+    get<T extends ResponseData>(url: string): Promise<void | T>
+}
 
-    get<T extends ResponseData>(url: string) {
-        return axios.get(url)
-            .then((response: AxiosResponse<Success>) => {
-                const payload = response && response.data;
-                return payload.data as T
-            })
-            .catch((e: AxiosError<Failure>) => {
-                return this.handleError(e)
-            })
+export class APIService<S extends string, E extends string> implements API<S, E> {
+
+    private readonly _commit: Commit;
+
+    constructor(commit: Commit) {
+        this._commit = commit
     }
 
-    postAndReturn<T extends ResponseData>(url: string, data: any) {
-        return axios.post(url, data)
-            .then((response: AxiosResponse<Success>) => {
-                const payload = response && response.data;
-                return payload.data as T
-            })
-            .catch((e: AxiosError<Failure>) => {
-               return this.handleError(e)
-            });
+    // appUrl will be set as a jest global during testing
+    private readonly _baseUrl = typeof appUrl !== "undefined" ? appUrl : "";
+
+    private _buildFullUrl = (url: string) => {
+        return this._baseUrl + url
+    };
+
+    private _ignoreErrors: Boolean = false;
+
+    static getFirstErrorFromFailure = (failure: Failure) => {
+        const firstError = failure.errors[0];
+        return firstError.detail ? firstError.detail : firstError.error;
+    };
+
+    private _onError: ((failure: Failure) => void) | null = null;
+
+    private _onSuccess: ((success: Success) => void) | null = null;
+
+    withError = (type: E) => {
+        this._onError = (failure: Failure) => {
+            this._commit({type: type, payload: APIService.getFirstErrorFromFailure(failure)});
+        };
+        return this
+    };
+
+    ignoreErrors = () => {
+        this._ignoreErrors = true;
+        return this;
+    };
+
+    withSuccess = (type: S) => {
+        this._onSuccess = (success: Success) => {
+            this._commit({type: type, payload: success.data});
+        };
+        return this;
+    };
+
+    private _handleAxiosResponse(promise: Promise<AxiosResponse>) {
+        return promise.then((response: AxiosResponse) => {
+            const success = response && response.data;
+            if (this._onSuccess) {
+                this._onSuccess(success);
+            }
+            return success;
```

Comment:
I've left this still returning the success object here, in case for any reason a consumer wants to handle the response themselves